File: ACTIVIDADES/Actividad N°5/Quick_Sort.py
import random
import time

# Algoritmo Quick Sort con pivote aleatorio
def quick_sort(arr, low, high, nivel=0):
    if low < high:
        pi = partition(arr, low, high)

        if nivel < 3:
            print(f"Partición nivel {nivel}: {arr[low:high+1][:10]}...")

        quick_sort(arr, low, pi - 1, nivel + 1)
        quick_sort(arr, pi + 1, high, nivel + 1)

# Partición con pivote aleatorio
def partition(arr, low, high):
    # Elegir índice aleatorio como pivote
    pivot_index = random.randint(low, high)
    arr[pivot_index], arr[high] = arr[high], arr[pivot_index]  # Mover pivote al final

    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

# Medir tiempo y mostrar
def probar_quick_sort(nombre, arreglo):
    print(f"\n--- {nombre.upper()} ---")
    copia = arreglo.copy()
    inicio = time.time()
    quick_sort(copia, 0, len(copia) - 1)
    fin = time.time()
    print(f"{nombre} - Tiempo: {fin - inicio:.6f} segundos")

# Arreglos
n = 1000
aleatorio = random.sample(range(n), n)
ordenado = list(range(n))
inverso = list(range(n - 1, -1, -1))

# Ejecutar
print("Quick Sort:")
probar_quick_sort("Aleatorio", aleatorio)
probar_quick_sort("Ordenado", ordenado)
probar_quick_sort("Inverso", inverso)

File: ACTIVIDADES/Actividad N°5/Quick Sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <random>
#include <string>

// Merge Sort recursivo con impresión de los primeros niveles
void merge_sort(std::vector<int>& arr, int nivel = 0) {
    if (arr.size() > 1) {
        int mid = arr.size() / 2;
        std::vector<int> L(arr.begin(), arr.begin() + mid);
        std::vector<int> R(arr.begin() + mid, arr.end());

        merge_sort(L, nivel + 1);
        merge_sort(R, nivel + 1);

        int i = 0, j = 0, k = 0;

        // Mezclar subarreglos ordenados
        while (i < L.size() && j < R.size()) {
            if (L[i] < R[j]) {
                arr[k++] = L[i++];
            } else {
                arr[k++] = R[j++];
            }
        }

        // Copiar lo que queda de L (si hay)
        while (i < L.size()) {
            arr[k++] = L[i++];
        }

        // Copiar lo que queda de R (si hay)
        while (j < R.size()) {
            arr[k++] = R[j++];
        }

        // Mostrar las fusiones de los primeros 3 niveles
        if (nivel < 3) {
            std::cout << "Fusion nivel " << nivel << ": ";
            for (size_t m = 0; m < std::min(arr.size(), size_t(10)); ++m) {
                std::cout << arr[m] << " ";
            }
            std::cout << "...\n";
        }
    }
}

// Función para probar el algoritmo y medir tiempo
void probar_merge_sort(const std::string& nombre, const std::vector<int>& arreglo) {
    std::cout << "\n--- " << nombre << " ---\n";
    std::vector<int> copia = arreglo; // Copia del arreglo original
    auto inicio = std::chrono::high_resolution_clock::now();
    merge_sort(copia);
    auto fin = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> duracion = fin - inicio;
    std::cout << nombre << " - Tiempo: " << duracion.count() << " segundos\n";
}

int main() {
    const int n = 1000;

    // Generar arreglo aleatorio
    std::vector<int> aleatorio(n);
    for (int i = 0; i < n; ++i) aleatorio[i] = i;
    std::random_device rd;
    std::mt19937 g(rd());
    std::shuffle(aleatorio.begin(), aleatorio.end(), g);

    // Arreglo ordenado
    std::vector<int> ordenado(n);
    for (int i = 0; i < n; ++i) ordenado[i] = i;

    // Arreglo inverso
    std::vector<int> inverso(n);
    for (int i = 0; i < n; ++i) inverso[i] = n - i;

    std::cout << "Merge Sort:\n";
    probar_merge_sort("Aleatorio", aleatorio);
    probar_merge_sort("Ordenado", ordenado);
    probar_merge_sort("Inverso", inverso);

    return 0;
}
